File: library/src/test/java/me/proxer/library/api/HeaderInterceptorTest.java
package me.proxer.library.api;

import me.proxer.library.BuildConfig;
import me.proxer.library.ProxerTest;
import me.proxer.library.util.ProxerUrls;
import okhttp3.Headers;
import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.mockwebserver.MockResponse;
import org.junit.jupiter.api.Test;

import java.io.IOException;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

/**
 * @author Ruben Gees
 */
class HeaderInterceptorTest extends ProxerTest {

    @Test
    void testApiKey() throws IOException, ProxerException, InterruptedException {
        server.enqueue(new MockResponse().setBody(fromResource("news.json")));
        api.notifications().news().build().execute();

        assertThat(server.takeRequest().getHeaders().get("proxer-api-key")).isEqualTo("mockKey");
    }

    @Test
    void testDefaultUserAgent() throws IOException, ProxerException, InterruptedException {
        server.enqueue(new MockResponse().setBody(fromResource("news.json")));
        api.notifications().news().build().execute();

        assertThat(server.takeRequest().getHeaders().get("User-Agent"))
                .isEqualTo("ProxerLibJava/" + BuildConfig.VERSION);
    }

    @Test
    void testCustomUserAgent() throws IOException, ProxerException, InterruptedException {
        api = constructApi()
                .userAgent("test")
                .build();

        server.enqueue(new MockResponse().setBody(fromResource("news.json")));
        api.notifications().news().build().execute();

        assertThat(server.takeRequest().getHeaders().get("User-Agent")).isEqualTo("test");
    }

    @Test
    void testCorrectHeadersForCdn() throws IOException, InterruptedException {
        server.enqueue(new MockResponse());
        api.client().newCall(new Request.Builder().url(ProxerUrls.cdnBase()
                .newBuilder()
                .addPathSegment("test")
                .build()).build()).execute();

        final Headers headers = server.takeRequest().getHeaders();

        assertThat(headers.get("proxer-api-key")).isNull();
        assertThat(headers.get("User-Agent")).startsWith("ProxerLibJava/");
    }

    @Test
    void testCorrectHeadersForStreamServer() throws IOException, InterruptedException {
        server.enqueue(new MockResponse());
        api.client().newCall(new Request.Builder().url("https://stream.proxer.me/test").build()).execute();

        final Headers headers = server.takeRequest().getHeaders();

        assertThat(headers.get("proxer-api-key")).isNull();
        assertThat(headers.get("User-Agent")).startsWith("ProxerLibJava/");
    }

    @Test
    void testCorrectHeadersForSpecificStreamServer() throws IOException, InterruptedException {
        server.enqueue(new MockResponse());
        api.client().newCall(new Request.Builder().url("https://s3-ps.proxer.me/test").build()).execute();

        final Headers headers = server.takeRequest().getHeaders();

        assertThat(headers.get("proxer-api-key")).isNull();
        assertThat(headers.get("User-Agent")).startsWith("ProxerLibJava/");
    }

    @Test
    void testCorrectHeadersForMangaServer() throws IOException, InterruptedException {
        server.enqueue(new MockResponse());
        api.client().newCall(new Request.Builder().url("https://manga0.proxer.me/test").build()).execute();

        final Headers headers = server.takeRequest().getHeaders();

        assertThat(headers.get("proxer-api-key")).isNull();
        assertThat(headers.get("User-Agent")).startsWith("ProxerLibJava/");
    }

    @Test
    void testOtherHostThrows() {
        HeaderInterceptor interceptor = new HeaderInterceptor("mock-key", "mock-user-agent");
        Interceptor.Chain chain = mock(Interceptor.Chain.class);

        when(chain.request()).thenReturn(new Request.Builder().url("https://example.com").build());

        assertThatExceptionOfType(IllegalArgumentException.class).isThrownBy(() -> interceptor.intercept(chain));
    }
}

File: library/src/test/kotlin/me/proxer/library/internal/interceptor/HeaderInterceptorTest.kt
package me.proxer.library.internal.interceptor

import io.mockk.every
import io.mockk.mockk
import io.mockk.slot
import me.proxer.library.BuildConfig.VERSION
import me.proxer.library.ProxerApi
import me.proxer.library.ProxerTest
import me.proxer.library.runRequest
import me.proxer.library.util.ProxerUrls
import okhttp3.Interceptor
import okhttp3.Request
import org.amshove.kluent.invoking
import org.amshove.kluent.shouldBe
import org.amshove.kluent.shouldBeNull
import org.amshove.kluent.shouldEqual
import org.amshove.kluent.shouldThrow
import org.junit.jupiter.api.Test

/**
 * @author Ruben Gees
 */
class HeaderInterceptorTest : ProxerTest() {

    private val interceptor = HeaderInterceptor("mock-key", "mock-user-agent")
    private val requestSlot = slot<Request>()

    private val chain = mockk<Interceptor.Chain>().apply {
        every { proceed(capture(requestSlot)) } returns mockk()
    }

    @Test
    fun testApiKey() {
        val (_, request) = server.runRequest("news.json") {
            api.notifications.news().build().execute()
        }

        request.headers["proxer-api-key"] shouldEqual "mock-key"
    }

    @Test
    fun testDefaultUserAgent() {
        val (_, request) = server.runRequest("news.json") {
            api.notifications.news().build().execute()
        }

        request.headers["User-Agent"] shouldEqual "ProxerLibJava/$VERSION"
    }

    @Test
    fun testApiKeyHeaderNotSetForOtherUrl() {
        every { chain.request() } returns Request.Builder().url(ProxerUrls.cdnBase).build()

        interceptor.intercept(chain)

        requestSlot.isCaptured shouldBe true
        requestSlot.captured.header("proxer-api-key").shouldBeNull()
    }

    @Test
    fun testCorrectHeadersForTestMode() {
        every { chain.request() } returns Request.Builder().url(ProxerUrls.apiBase).build()

        HeaderInterceptor(ProxerApi.TEST_KEY, "mock-user-agent").intercept(chain)

        requestSlot.isCaptured shouldBe true
        requestSlot.captured.header("proxer-api-key").shouldBeNull()
        requestSlot.captured.header("proxer-api-testmode") shouldEqual "1"
    }

    @Test
    fun testTestModeHeaderNotSerForOtherUrl() {
        every { chain.request() } returns Request.Builder().url(ProxerUrls.cdnBase).build()

        HeaderInterceptor(ProxerApi.TEST_KEY, "mock-user-agent").intercept(chain)

        requestSlot.isCaptured shouldBe true
        requestSlot.captured.header("proxer-api-testmode").shouldBeNull()
    }

    @Test
    fun testCorrectHeadersForCdn() {
        every { chain.request() } returns Request.Builder().url(ProxerUrls.cdnBase).build()

        interceptor.intercept(chain)

        requestSlot.isCaptured shouldBe true
        requestSlot.captured.header("User-Agent").toString() shouldEqual "mock-user-agent"
    }

    @Test
    fun testCorrectHeadersForStreamServer() {
        every { chain.request() } returns Request.Builder().url(ProxerUrls.streamBase).build()

        interceptor.intercept(chain)

        requestSlot.isCaptured shouldBe true
        requestSlot.captured.header("User-Agent").toString() shouldEqual "mock-user-agent"
    }

    @Test
    fun testCorrectHeadersForSpecificStreamServer() {
        every { chain.request() } returns Request.Builder().url("https://s3-ps.proxer.me").build()

        interceptor.intercept(chain)

        requestSlot.isCaptured shouldBe true
        requestSlot.captured.header("User-Agent").toString() shouldEqual "mock-user-agent"
    }

    @Test
    fun testCorrectHeadersForMangaServer() {
        every { chain.request() } returns Request.Builder().url("https://manga0.proxer.me").build()

        interceptor.intercept(chain)

        requestSlot.isCaptured shouldBe true
        requestSlot.captured.header("User-Agent").toString() shouldEqual "mock-user-agent"
    }

    @Test
    fun testCorrectHeadersForProxy() {
        every { chain.request() } returns Request.Builder().url("https://proxy.proxer.me").build()

        interceptor.intercept(chain)

        requestSlot.isCaptured shouldBe true
        requestSlot.captured.header("User-Agent").toString() shouldEqual "mock-user-agent"
    }

    @Test
    fun testOtherHostThrows() {
        every { chain.request() } returns Request.Builder().url("https://example.com").build()

        invoking { interceptor.intercept(chain) } shouldThrow IllegalArgumentException::class
    }
}
